Add changelog for beta.18 (#278)
mapbox/mapbox-maps-android: CHANGELOG.md

Mapbox welcomes participation and contributions from everyone.

# 10.0.0-beta.18 - April 22, 2021
## Breaking changes ⚠️

> * ResourceOptions contains tile store instance (TileStore API). Tile store usage is enabled by default, 'ResourceOptions.tileStoreEnabled' flag is introduced to disable it.

Reviewer: ```suggestion
```

> * ResourceOptions contains tile store instance (TileStore API). Tile store usage is enabled by default, 'ResourceOptions.tileStoreEnabled' flag is introduced to disable it.

Reviewer 2: ```suggestion
* Change `tileStorePath` to `tileStore` inside the ResourceOptions. Tile store usage is enabled by default to support the offline feature, 'ResourceOptions.tileStoreEnabled' flag is introduced to disable it.
```

* Rename MapView plugin extension functions. ([#272](https://github.com/mapbox/mapbox-maps-android/pull/272))
  - mapView.getAnnotationPlugin() -> mapView.annotations
  - mapView.getGesturesPlugin() -> mapView.gestures
  - mapView.getOverlayPlugin() -> mapView.overlay() // using function here because of experimental annotation
  - mapView.getLocationComponentPlugin() -> mapView.location
  - mapView.getCameraAnimationsPlugin() -> mapView.camera
  - mapView.getAttributionPlugin() -> mapView.attribution
  - mapView.getCompassPlugin() -> mapView.compass
  - mapView.getLogoPlugin() -> mapView.logo
  - mapView.getScaleBarPlugin() -> mapView.scalebar
* Remove deprecated location plugin ([#276](https://github.com/mapbox/mapbox-maps-android/pull/276))
* Add feature sdk initialisation ([#269](https://github.com/mapbox/mapbox-maps-android/pull/269))

>   - Load the Mapbox Street style by default if user doesn't load any style before the onStart lifecycle event.

Reviewer 3: @Chaoba we are now having that entry both in breaking changes and features, is this intended?

  - Introduce `MapInitOptions` to replace MapboxMapOptions. When adding a map programmatically, client codes can be written in a following way
  ```
   // set map options
    val mapOptions = MapOptions.Builder()
      .constrainMode(ConstrainMode.HEIGHT_ONLY)
      .glyphsRasterizationOptions(
        GlyphsRasterizationOptions.Builder()
          .rasterizationMode(GlyphsRasterizationMode.IDEOGRAPHS_RASTERIZED_LOCALLY)
          .build()
      )
      .build()

    // set token and cache size for this particular map view
    val resourceOptions = ResourceOptions.Builder()
      .accessToken(getString(R.string.mapbox_access_token))
      .cacheSize(75_000L)
      .build()

    // set initial camera position
    val initialCameraOptions = CameraOptions.Builder()
      .center(Point.fromLngLat(-122.4194, 37.7749))
      .zoom(9.0)
      .build()

    val mapboxMapOptions =
      MapInitOptions(this, resourceOptions, mapOptions, initialCameraOptions, true)
    // create view programmatically
    customMapView = MapView(this, mapboxMapOptions)

>   ```

Reviewer 4: Should we leave this example code to the Github release notes and the new guide?

Reviewer 2: I think it doesn't hurt to keep them here, we can reorganise it later if needed.

>   - Introduce `MapInitOptions` to replace MapboxMapOptions. When adding a map programmatically, client codes can be written in a following way
>   ```
>    // set map options
>     val mapOptions = MapOptions.Builder()
>       .constrainMode(ConstrainMode.HEIGHT_ONLY)
>       .glyphsRasterizationOptions(
>         GlyphsRasterizationOptions.Builder()
>           .rasterizationMode(GlyphsRasterizationMode.IDEOGRAPHS_RASTERIZED_LOCALLY)
>           .build()
>       )
>       .build()
> 
>     // set token and cache size for this particular map view
>     val resourceOptions = ResourceOptions.Builder()
>       .accessToken(getString(R.string.mapbox_access_token))
>       .cacheSize(75_000L)
>       .build()
> 
>     // set initial camera position
>     val initialCameraOptions = CameraOptions.Builder()
>       .center(Point.fromLngLat(-122.4194, 37.7749))
>       .zoom(9.0)
>       .build()
> 
>     val mapboxMapOptions =
>       MapInitOptions(this, resourceOptions, mapOptions, initialCameraOptions, true)
>     // create view programmatically
>     customMapView = MapView(this, mapboxMapOptions)
>   ```

Reviewer: ```suggestion
  - Introduce `MapInitOptions` to replace MapboxMapOptions. 
```

## Features ✨ and improvements 🏁
* High-level animations return cancelable interface ([#262](https://github.com/mapbox/mapbox-maps-android/pull/262))
* Introduce OfflineManager API that manages style packs and produces tileset descriptors for the tile store.
  - By default, users may download up to 250MB of data for offline use without incurring additional charges. This limit is subject to change during the beta.

>   - The new API replaces the deprecated OfflineRegionManager API. The OfflineManager API can be used to create offline style packs that contain style data, such as: style definition, sprites, fonts and other resources. Tileset descriptors created by the OfflineManager API are used to create tile packs via TileStore API. Mobile maps SDKs use tile packs for rendering map content. Example of how to use API is presented in a following snippet (pseudocode):
>  ```
>   // 1. Get an instance of OfflineManager.
>   val offlineManager = OfflineManager(map.getResourceOptions())
> 
>   // 2. Create a tile region with tiles for Streets
>   TilesetDescriptorOptions tilesetOptions(
>       "mapbox://styles/mapbox/streets-v11",
>       0, // min zoom
>       5, // max zoom
>   );
> 
>   // Creation of a tileset descriptor WILL implicitly create a style package for provided StylePackLoadOptions.
>   tilesetOptions.stylePackOptions = StylePackLoadOptions(GlyphsRasterizationMode.AllGlyphsRasterizedLocally);
>   val descriptor = offlineManager.createTilesetDescriptor(tilesetOptions);
> 
>   val tileStore = TileStore.getInstance();
>   TileRegionLoadOptions tileRegionLoadOptions(geometry, [descriptor]);
>   tileStore.loadTileRegion("my_region", tileRegionLoadOptions,
>     onProgressCallback(result) {
>       // Handle TileRegionLoadProgress result
>     },
>     onFinishedCallback(result) {
>       if (result) {
>         // Handle TileRegion result
>       } else {
>         // Handle TileRegionError
>       }
>     }
>   );
>   ```

Reviewer: ```suggestion
  - The new API replaces the deprecated OfflineRegionManager API. The OfflineManager API can be used to create offline style packs that contain style data, such as: style definition, sprites, fonts and other resources. Tileset descriptors created by the OfflineManager API are used to create tile packs via TileStore API. Mobile maps SDKs use tile packs for rendering map content. 
```

>   ```

Reviewer 4: Same here.

      .zoom(9.0)
      .build()

    val mapboxMapOptions =
      MapInitOptions(this, resourceOptions, mapOptions, initialCameraOptions, true)
    // create view programmatically
    customMapView = MapView(this, mapboxMapOptions)
  ```
* Expose source, source layer id and feature state in query*Features APIs' results.
  The change modifies arguments signature of a QueryFeaturesCallback from
  expected<array<feature>, string> to expected<array<QueriedFeature>, string>. Newly introduced QueriedFeature struct contains queried feature as well as other fields, such as: source, sourceLayer and feature state represented by state field.
  Client code has to be updated in a following way (pseudocode):
  ```
  // Before
  map.queryRenderedFeatures(callback(features) {
    for (feature : features) {
      val id = feature.id;
    }
  });
  
  // After
  map.queryRenderedFeatures(callback(queriedFeatures) {
    for (queriedFeature : queriedFeatures) {
      val id = queriedFeature.feature.id;
    }
  });
  ```
## Features ✨ and improvements 🏁
* [animation] High-level animations return cancelable interface ([#262](https://github.com/mapbox/mapbox-maps-android/pull/262))

> * Enable identical code folding and -lto-O3 optimizations to reduce size of Android Mobile SDK 

Author: Oh, yes, copy the wrong release

* Integrate map engine with NetworkConnectivity API. Map will respect setMapboxStackConnected setting and decide whether http requests can be made.
* Tile Store
  - Add index for group <-> tiles to speed up eviction queries
  - Don't return cached responses for 401/403/Unauthorized requests
  - The invoked async methods complete even if the client code does not keep the tile store instance any more.
## Bug fixes 🐞
* Keep CompassPlugin enabled/disabled state after other properties update ([#252](https://github.com/mapbox/mapbox-maps-android/pull/252))

> * Fix distance expression parsing for geometries containing over 20k points.

Reviewer 3: was in prev release

Mapbox welcomes participation and contributions from everyone.

# 10.0.0-beta.18 - April 22, 2021

> ## Breaking changes ⚠️

Reviewer 2: should also include https://github.com/mapbox/mapbox-maps-android/pull/272

Reviewer 2: and https://github.com/mapbox/mapbox-maps-android/pull/276

Reviewer 2: and also https://github.com/mapbox/mapbox-maps-android/pull/228

* ResourceOptions contains tile store instance (TileStore API). Tile store usage is enabled by default, the resource option 'tileStoreEnabled' flag is introduced to disable it.

> * Add feature sdk initialisation ([#269](https://github.com/mapbox/mapbox-maps-android/pull/269))

Reviewer 2: We could rephrase this entry to be more concrete, something like:

Introduce the MapInitOptions to replace MapboxMapOptions.
Introduce CredentialsManager to help manage Mapbox access tokens.
Load the Mapbox Street style by default if user doesn't load any style before the onStart lifecycle event.

etc.

Reviewer 2: or sorry just noticed the sections below. 🤦

Reviewer 2: still the renaming of MapboxMapOptions to MapInitOptions is still worth mentioning here.
And the default style is used now.

 ```
  // 1. Get an instance of OfflineManager.
  val offlineManager = OfflineManager(map.getResourceOptions())

  // 2. Create a tile region with tiles for Streets
  TilesetDescriptorOptions tilesetOptions(
      "mapbox://styles/mapbox/streets-v11",
      0, // min zoom
      5, // max zoom
  );

  // Creation of a tileset descriptor WILL implicitly create a style package for provided StylePackLoadOptions.
  tilesetOptions.stylePackOptions = StylePackLoadOptions(GlyphsRasterizationMode.AllGlyphsRasterizedLocally);
  val descriptor = offlineManager.createTilesetDescriptor(tilesetOptions);

  val tileStore = TileStore.getInstance();
  TileRegionLoadOptions tileRegionLoadOptions(geometry, [descriptor]);
  tileStore.loadTileRegion("my_region", tileRegionLoadOptions,
    onProgressCallback(result) {
      // Handle TileRegionLoadProgress result
    },
    onFinishedCallback(result) {
      if (result) {
        // Handle TileRegion result
      } else {
        // Handle TileRegionError
      }
    }
  );
  ```

> * Tile Store

Author: Do you think it's worth mentioning these changes?

Reviewer 2: I don't think these are very customer-facing, especially we just introduced the TileStore to be used in the Maps side with the Offline manager in this release.

  - Add index for group <-> tiles to speed up eviction queries
  - Don't return cached responses for 401/403/Unauthorized requests
  - The invoked async methods complete even if the client code does not keep the tile store instance any more.
## Bug fixes 🐞

> * Keep CompassPlugin enabled/disabled state after other properties update ([#252](https://github.com/mapbox/mapbox-maps-android/pull/252))

Reviewer 2: should also include  https://github.com/mapbox/mapbox-maps-android/pull/273

Reviewer 3: should also include https://github.com/mapbox/mapbox-maps-android/pull/279

Reviewer 2: and https://github.com/mapbox/mapbox-maps-android/pull/228

  );

  // Creation of a tileset descriptor WILL implicitly create a style package for provided StylePackLoadOptions.
  tilesetOptions.stylePackOptions = StylePackLoadOptions(GlyphsRasterizationMode.AllGlyphsRasterizedLocally);
  val descriptor = offlineManager.createTilesetDescriptor(tilesetOptions);

  val tileStore = TileStore.getInstance();
  TileRegionLoadOptions tileRegionLoadOptions(geometry, [descriptor]);
  tileStore.loadTileRegion("my_region", tileRegionLoadOptions,
    onProgressCallback(result) {
      // Handle TileRegionLoadProgress result
    },
    onFinishedCallback(result) {
      if (result) {
        // Handle TileRegion result
      } else {
        // Handle TileRegionError
      }
    }
  );
  ```

## Bug fixes 🐞
* Keep CompassPlugin enabled/disabled state after other properties update ([#252](https://github.com/mapbox/mapbox-maps-android/pull/252))
* Fix disabling logo in xml. ([#273](https://github.com/mapbox/mapbox-maps-android/pull/273))
* Introduce StyleInterface that include the current display's pixel ratio, and fix Style#addImage to take the correct pixel ratio from display.  ([#228](https://github.com/mapbox/mapbox-maps-android/pull/228))

> * Properly reset anchor after some gestures ([#279](https://github.com/mapbox/mapbox-maps-android/pull/279))

Reviewer 3: Let's add this small fix https://github.com/mapbox/mapbox-maps-android/pull/280 and also https://github.com/mapbox/mapbox-maps-android/pull/259 as an initial example how to use offline api

Mapbox welcomes participation and contributions from everyone.

# 10.0.0-beta.18 - April 22, 2021
## Breaking changes ⚠️
* ResourceOptions contains tile store instance (TileStore API). Tile store usage is enabled by default, the resource option 'tileStoreEnabled' flag is introduced to disable it.

> * Rename MapView plugin extension functions. ([#272](https://github.com/mapbox/mapbox-maps-android/pull/272))

Reviewer 2: can we be more specific here, and list the previous name and after name?

Mapbox welcomes participation and contributions from everyone.

# 10.0.0-beta.18 - April 22, 2021
## Breaking changes ⚠️

> * ResourceOptions contains tile store instance (TileStore API). Tile store usage is enabled by default, the resource option 'tileStoreEnabled' flag is introduced to disable it.

Reviewer 3: Can we specify `ResourceOptions.tileStoreEnabled`?

* Rename MapView plugin extension functions. ([#272](https://github.com/mapbox/mapbox-maps-android/pull/272))
  - mapView.getAnnotationPlugin() -> mapView.annotations

>   - mapView.getGesturesPlugin() ->   - mapView.gestures

Reviewer 2: ```suggestion
  - mapView.getGesturesPlugin() ->   mapView.gestures
```

  - mapView.getOverlayPlugin() -> mapView.overlay() // using function here because of experimental annotation
  - mapView.getLocationComponentPlugin() -> mapView.location
  - mapView.getCameraAnimationsPlugin() -> mapView.camera
  - mapView.getAttributionPlugin() -> mapView.attribution
  - mapView.getCompassPlugin() -> mapView.compass
  - mapView.getLogoPlugin() -> mapView.logo
  - mapView.getScaleBarPlugin() -> mapView.scalebar
* Remove deprecated location plugin ([#276](https://github.com/mapbox/mapbox-maps-android/pull/276))
* Add feature sdk initialisation ([#269](https://github.com/mapbox/mapbox-maps-android/pull/269))

>   - Load the Mapbox Street style by default if user doesn't load any style before the onStart lifecycle event.

Reviewer 3: Is it a breaking change?

Author: It's included by sdk initilisation.

Reviewer 3: Technically yes, but for end user those 2 things are not connected, loading style by default for me seems to be under `features and improvements`.

Author: That makes sense, let's move it.

Reviewer 2: Well it's also a breaking change in terms of behaviour.

Reviewer 3: @pengdev following that logic fixing every bug or feature could be treated as breaking change 😄 
imho breaking change for user is basically breaking an API - so that using version X project builds fine while using X + 1 will fail to compile.
